File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        ArrayList<String> list=new ArrayList();
        generate(list,"",0,0,n);
        return list;
    }
    static void generate(ArrayList list,String str,int open,int close,int n) {
		if(str.length()==2*n) {
			list.add(str);
			return;
		}
		if(open<n) {
			generate(list,str+"(",open+1,close,n);
		}
		if(close<open) {
			generate(list,str+")",open,close+1,n);
		}
	}
}

File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    void backtrack(vector<string>&v,int n,int oc,int cc,string s)
    {
        if(oc==n && cc==n)
        {
          v.push_back(s);
          return;
        }
        if(oc<n)
        {
        backtrack(v,n,oc+1,cc,s+"(");
        }
        if(cc<oc)
        {
        backtrack(v,n,oc,cc+1,s+")");
        }
         
    }
    vector<string> generateParenthesis(int n) {
        vector<string> v;
        string s;
        backtrack(v,n,0,0,"");
        return v;
        
    }
};